File: dissertation/model/game.py
"""Functions for creating communication games/fitness functions."""

import pyximport; pyximport.install()
from itertools import izip

from cython_funcs import comm_success


def communication_scenario_factory(reference_costs, ambiguous_reference_cost, success_points):
    """Create a function which will evaluate the communicative success of two agents."""
    def comm(player, partner):
        """
        Determine the cost incurred and points awarded when player attempts to
        communicate all meanings to their partner.
        """
        partner_sum = float(sum(partner))
        return sum(
            comm_success(
                cost,
                ambiguous_reference_cost,
                success_points,
                player_ambiguity_probability,
                partner_ambiguity_probability,
                partner_sum
            )
            for cost, player_ambiguity_probability, partner_ambiguity_probability in
            izip(reference_costs, player, partner)
        )
    return comm


def game_factory(comm_func):
    """Create a fitness function which evalutates an agent's fitness within its group."""
    def game(player, group):
        """
        Given group positions and a player position, run a communication game and
        return the player's score.
        """
        return sum(
            comm_func(player, partner) + comm_func(partner, player)
            for partner in group
            if partner is not player
        )
    return game


two_item_game = game_factory(
    communication_scenario_factory(
        reference_costs=(-1., -2.),
        ambiguous_reference_cost=-1.25,
        success_points=1.
    )
)

File: dissertation/model/game.hs
import Data.List


single_comm_success :: Float -> Float -> Float -> Float -> Float -> Float -> Float
single_comm_success ambiguous_reference_cost success_points partner_sum cost player_prob partner_prob
	| partner_prob > 0.0 = ambiguous_cost + ambiguous_reward + unambiguous_points
	| otherwise = ambiguous_cost + unambiguous_points
	where
		ambiguous_cost = ambiguous_reference_cost * player_prob
		ambiguous_reward = success_points * player_prob * (partner_prob / partner_sum)
		unambiguous_points = ((cost + success_points) * (1.0 - player_prob))

comm_success :: [Float] -> Float -> Float -> [Float] -> [Float] -> Float
comm_success costs ambiguous_reference_cost success_points player partner =
	sum $ zipWith3 item_comm_success costs player partner
	where
		partner_sum = sum partner
		item_comm_success = single_comm_success ambiguous_reference_cost success_points partner_sum

game :: ([Float] -> [Float] -> Float) -> [Float] -> [[Float]] -> Float -- Maybe have this take particles, have position implement eq?
game comm_func player [] = 0.0
game comm_func player group = sum $ map (pair_success player) $ delete player group
    where
    	pair_success player partner = comm_func player partner + comm_func partner player

main = do
	putStrLn $ show $ comm_success [0] 0 1 [1] [1]
	putStrLn $ show $ comm_success [0] 0 1 [1] [0]
	putStrLn $ show $ comm_success [0] 0 1 [1] [0.5]
	putStrLn $ show $ comm_success [0] 0 1 [1] [1]
	putStrLn $ show $ comm_success [0] 0 1 [0] [1]
	putStrLn $ show $ comm_success [0] 0 1 [0.5] [1]
	return $ (game $ comm_success [0] 0 1) [1] [[0.99], [0], [0.5]]